### PR TITLE
[None][perf] DSv4 follow-up: sparse attention and model defaults

### PR DESCRIPTION
**Pull request by lfr-0531**
@coderabbitai summary

## Description

This PR ports the PR-3 slice of the DeepSeek V4 post-`a7728aa` work onto the current GitHub `main` branch.

It focuses on DSv4 sparse attention execution and the tightly-coupled custom op/model defaults:

- Wires fused FP8 Q quantization and `quant_scale_qkv` through the TRT-LLM THOP attention path while preserving the newer `main` attention API shape.
- Adds/updates DSv4 QNorm, inverse-RoPE FP8 quantization, indexer Top-K, compressor, and FP8 quantization kernels/op bindings.
- Updates DSv4 sparse attention/model defaults, including FP4 indexer K-cache defaults and KV cache manager v2 default handling.
- Expands DSv4 sparse/custom-op/unit coverage and DSv4 Flash accuracy references.

Conflict handling was documented locally during branch preparation and is intentionally not included in this PR because `tmp/` contents should not be submitted.

The branch was prepared in a fresh worktree and has been rebased onto GitHub `main` at `e9016273e8e87ee4998425e14ab37ca4f5e59e11`.

## Test Coverage

Build/install:

- PASS: `python3 ./scripts/build_wheel.py --trt_root /usr/local/tensorrt --benchmarks --use_ccache --cuda_architectures "90-real;100-real" --configure_cmake`
  - Native CMake build completed.
  - Wheel packaging initially hit a local Python environment mismatch: PyPI torch in `.venv-3.12` conflicted with the container NVIDIA torch.
  - Reconfigured/rebuilt `th_common` and `bindings` against `/usr/local/lib/python3.12/dist-packages/torch` and synced the rebuilt `.so` files into the source package path.
- PASS: import verification from this worktree:
  - torch realpath: `/usr/local/lib/python3.12/dist-packages/torch/__init__.py`
  - `tensorrt_llm`: this PR worktree
  - `tensorrt_llm.bindings`: this PR worktree rebuilt binding
  - `modeling_deepseekv4.py`: this PR worktree

Static checks:

- PASS: `git diff --check github/main...HEAD`
- PASS: `rg -n '^(<<<<<<<|>>>>>>>)'`
- PASS: PR-3 forbidden-scope check found no matches for unrelated MoE/server/disaggregated/FMHA test-list paths.

GPU pre-check:

- Checked `nvidia-smi` before GPU tests and selected an idle GPU at the time.

Completed tests:

- PASS: `CUDA_VISIBLE_DEVICES=1 timeout 20m .venv-3.12/bin/python -m pytest tests/unittest/_torch/custom_ops/test_deepseek_v4_q_norm.py -q --tb=short`
  - `81 passed`
- PASS: `CUDA_VISIBLE_DEVICES=1 timeout 20m .venv-3.12/bin/python -m pytest tests/unittest/_torch/custom_ops/test_fused_inv_rope_fp8_quant.py -q --tb=short`
  - `128 passed`
- PASS: `CUDA_VISIBLE_DEVICES=1 timeout 20m .venv-3.12/bin/python -m pytest tests/unittest/_torch/attention/sparse/deepseek_v4 -q --tb=short`
  - `213 passed, 22 skipped`
- PASS: `CUDA_VISIBLE_DEVICES=1 timeout 20m .venv-3.12/bin/python -m pytest tests/unittest/_torch/thop/parallel/test_fp8_quantize.py -q --tb=short`
  - `147 passed, 9 skipped`

Partial / pending tests:

- PARTIAL: `CUDA_VISIBLE_DEVICES=1 timeout 20m .venv-3.12/bin/python -m pytest tests/unittest/_torch/attention/sparse -q --tb=short`
  - Reached 43% before the 20 minute timeout.
  - One failure marker appeared before timeout, but the full failure summary was not emitted before timeout.
  - The PR-3-focused `tests/unittest/_torch/attention/sparse/deepseek_v4` subset passed afterward.
- PENDING: `tests/unittest/_torch/thop/parallel/test_indexer_topk.py`
  - Started and was making progress, but the machine/GPU became shared again; stopped before completion to avoid occupying GPUs.
- PENDING: `tests/unittest/_torch/modeling/test_modeling_deepseekv4.py`
- NOT RUN: targeted DSv4 Flash accuracy integration coverage from `tests/integration/defs/accuracy/test_llm_api_pytorch.py`
  - `LLM_MODELS_ROOT` is not set in this environment, so model-weight-dependent accuracy tests need to run in CI or a configured model-weight environment.

## PR Checklist

- [x] I have added tests to cover my changes.
- [x] I have updated documentation when relevant.
- [x] I have added the required copyright header to new files.
- [x] I have read the contribution guidelines.

## GitHub Bot Help

See the repository bot help for available CI trigger commands.

### CODE REVIEW

**Comment by longcheng-nv** [comment]
While reviewing this PR I noticed a duplicate method definition in `tensorrt_llm/_torch/attention_backend/sparse/dsa.py`: `DSAtrtllmAttentionMetadata._create_radix_aux_buffers` is defined twice (around L878 and L1078). The two bodies are functionally identical (same shapes/dtypes), so Python silently keeps only the second one and the first (better-documented) definition becomes dead code. Looks like a merge/fold artifact from carrying over the radix-aux scratch fix.

Suggested cleanup: drop the redundant second definition and keep the documented one — no behavior change. Patch (applies cleanly on the current PR head):

```diff
@@ class DSAtrtllmAttentionMetadata(TrtllmAttentionMetadata):
         # Create expanded buffers for MTP support
         self.create_expanded_buffers(capture_graph=capture_graph)
 
-    def _create_radix_aux_buffers(self, capture_graph=False):
-        """Create persistent scratch for the fp32 radix split-work TopK path."""
-        max_blocks_per_row = 10
-        max_gen_tokens = self.max_num_sequences * (1 + self.max_draft_tokens)
-        self.radix_aux_indices = self.get_empty(
-            self.cuda_graph_buffers,
-            (max_gen_tokens, max_blocks_per_row, self.num_sparse_topk),
-            cache_name="radix_aux_indices",
-            dtype=torch.int32,
-            capture_graph=capture_graph,
-        )
-        self.radix_aux_logits = self.get_empty(
-            self.cuda_graph_buffers,
-            (max_gen_tokens, max_blocks_per_row, self.num_sparse_topk),
-            cache_name="radix_aux_logits",
-            dtype=torch.float32,
-            capture_graph=capture_graph,
-        )
-
     def _create_kv_lens_2d_buffer(self, capture_graph=False):
```

**Review by mingyangHao — APPROVED** [review]
LGTM

**Review by jieli-matrix — APPROVED** [review]

**Comment by lfr-0531** [comment]
Hit an IMA issue in DSV4 tests. We'll need to wait for https://github.com/NVIDIA/TensorRT-LLM/pull/15633 to land first.

File: tensorrt_llm/_torch/custom_ops/triton_fused_inv_rope_fp8_quant.py [hunk 1]
```diff
@@ -1,169 +1,34 @@
...
-# when `use_cute_dsl_blockscaling_bmm` is enabled in `extra_llm_api_options` on
-# SM100; falls back to the legacy `mla_rope_inplace + permute102_quant` pair
-# otherwise.
 """
-Fused inverse-RoPE + 1x128 block-scaled FP8 quantize, for DeepSeek-V4 ``o_a_proj``.
-
-Replaces the (``mla_rope_inplace`` -> ``fp8_batched_quantize_1x128_permute102``)
-two-kernel pair in ``_deepseek_v4_o_proj``. The output is consumed verbatim by
-``cute_dsl_fp8_bmm_blackwell``.
+``trtllm::fused_inv_rope_fp8_quant_vllm_port`` — fake/meta registration.
```
Comment (liji-nv) [inline tensorrt_llm/_torch/custom_ops/triton_fused_inv_rope_fp8_quant.py]: Better move to  cpp_custom_ops.py

**Review by xinhe-nv — APPROVED** [review]

**Review by JunyiXu-nv — APPROVED** [review]
LGTM for llmapi side

**Review by nv-guomingz — APPROVED** [review]

File: tensorrt_llm/_torch/custom_ops/triton_fused_inv_rope_fp8_quant.py (hunk 1 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/custom_ops/triton_fused_inv_rope_fp8_quant.py]: Fixed in https://github.com/NVIDIA/TensorRT-LLM/pull/15717/changes/e291925cc60214a0ae57ae28735a3772e46ae9e3. Thanks, Jin!

**Review by longcheng-nv — APPROVED** [review]
LGTM. CI (blossom-ci L0) green on this head, DCO/pre-commit pass; reviewed kernel/op wiring and test coverage — new FP8 Q-quant / inverse-RoPE quant / FP4 indexer K-cache defaults are reference-tested and SM-gated appropriately.

**Review by juney-nvidia — APPROVED** [review]
Approved to unblock the subsequent PRs firstly.
